services/src/main/java/org/keycloak/protocol/oidc/endpoints/UserInfoEndpoint.java

```diff
@@ -225,6 +226,24 @@ private Response issueUserInfo(String tokenString) {
         // any attempt to customize the value of this field should be done through a different claim
         claims.put("sub", userInfo.getSubject());
 
+        if (userInfo.getRealmAccess() != null) {
+            Map<String, Set<String>> realmAccess = new HashMap<>();
+            realmAccess.put("roles", userInfo.getRealmAccess().getRoles());
+            claims.put("realm_access", realmAccess);
+        }
+
+        if (userInfo.getResourceAccess() != null && !userInfo.getResourceAccess().isEmpty()) {
+            Map<String, Map<String, Set<String>>> resourceAccessMap = new HashMap<>();
+
+            for (Map.Entry<String, AccessToken.Access> resourceAccessMapEntry : userInfo.getResourceAccess()
+                    .entrySet()) {
+                Map<String, Set<String>> resourceAccess = new HashMap<>();
+                resourceAccess.put("roles", resourceAccessMapEntry.getValue().getRoles());
+                resourceAccessMap.put(resourceAccessMapEntry.getKey(), resourceAccess);
+            }
+            claims.put("resource_access", resourceAccessMap);
+        }
```

Comment:
Looks like a workaround to me and not a proper fix. Roles should be added by protocol mappers and not hardcoded here.

Reply:
The current mechanism is the following:

Some lines above in https://github.com/keycloak/keycloak/blob/3186f1b5a9103383b4f4d8b6f595449055c44b5b/services/src/main/java/org/keycloak/protocol/oidc/endpoints/UserInfoEndpoint.java#L219
the protocol mappers are used to create an access token with all the information that should be available in the user info endpoint. This is done in a generic way that applies to the access token, id token and user info.
If the default claim name for realm roles/client roles are not modified the roles are stored in a separate field in the access token, see https://github.com/keycloak/keycloak/blob/3186f1b5a9103383b4f4d8b6f595449055c44b5b/services/src/main/java/org/keycloak/protocol/oidc/mappers/AbstractUserRoleMappingMapper.java#L139
Since only `otherClaims` field is used to put info to the user info endpoint the bug occurs.

Which way would you suggest to properly fix this bug? I see the following options

1. Use the `realmAccess` and `resourceAccess` properties in the access token that the protocol mappers provided. Just fetch the information of them and put it in the user info right here. That's the current version of the PR.
2. Add the information to the protocol mappers that they should map the roles in a "user info endpoint" way. In this case just put the roles in `otherClaims` field of the token.
3. Don't use a special field for `realmAccess` and `resourceAccess` anymore such that the values are always provided in `otherClaims`.